feat(linter/plugins): allow JS plugins to access `globals`

lilnasy opened the pull request:
- This PR implements a naive, inefficient solution to #16359.
- To establish current and expected behavior, this PR serializes rust-resolved and -overridden globals once for every file.

## Future work
- [ ] Add test for globals from the `env` field. Implement support if not currently supported.
- [ ] Optimization: we want to serialize `globals` for all files only once at the start of the lint run.

## AI Usage
- This PR was drafted by Claude Sonnet 4.5 in Plan + Agent mode, then manually reviewed and iterated on.

lilnasy committed:
feat(linter/plugins): allow JS plugins to access `globals`

lilnasy committed:
`to_string_lossy()` does not exist anymore?

overlookmotel commented on apps/oxlint/src-js/plugins/lint.ts:
```diff
@@ -142,6 +148,9 @@ export function lintFileImpl(
   // Pass settings JSON to context module
   setSettingsForFile(settingsJSON);
 
+  // Pass globals JSON to context module
```
```suggestion
  // Pass globals JSON to globals module
```

The comment above for settings was also inaccurate. Fixed that in #16559.

overlookmotel reviewed (commented):
I know you've said this is just an experiment, but I think we should merge it!

Yes, it's inefficient to serialize globals for every file (ditto settings). But at least it works! Much better than just not supporting it at all, which is what we have now.

A few comments below - none major.

---

We also support an `env` option in config: https://oxc.rs/docs/guide/usage/linter/config.html#configuration-file-format

I'm not sure globals from `env` already get merged into `OxlintGlobals` on Rust side, in which case it'll work already after this PR.

If so, we should add a test for that. If not, we should add support for `env` (simplest way would be to just apply `env` to `OxlintGlobals` on Rust side, before serializing to JSON). Either way, I suggest leaving that to a follow-up PR.

lilnasy commented:
@overlookmotel Makes sense. I'll address the review comments tonight and add todo comments for performance.

lilnasy committed:
suggestions from code review

lilnasy committed:
style: import with ts extension

lilnasy commented on apps/oxlint/test/fixtures/globals/files/nested/index.js:
```diff
@@ -0,0 +1 @@
+let x;
```
It's a test. Leave it alone.

lilnasy commented:
Addressed all review comments. CI is failing because spell check flaked.

lilnasy commented on crates/oxc_linter/src/config/globals.rs:
```diff
@@ -61,6 +61,7 @@ impl OxlintGlobals {
 #[serde(rename_all = "lowercase")]
 pub enum GlobalValue {
     Readonly,
+    // TODO: #[serde(rename = "writable")] for ESLint compatibility
```
This can be done in a separate, smaller PR focused on making sure there are no breaking changes when we add the alias to the deserializer.

overlookmotel committed:
style: formatting

I wish `oxfmt` would do this automatically for us. It will, but not implemented yet...

overlookmotel committed:
style: move comment

overlookmotel committed:
add comments

overlookmotel committed:
add debug assert

overlookmotel committed:
`Object.freeze` as top-level var

Tiny optimization. Avoids property lookup on each call.

overlookmotel committed:
avoid `@ts-expect-error`

overlookmotel committed:
style: shorten code

overlookmotel reviewed (approved):
Excellent!

Pushed a few commits with my usual style nits, but made no substantive changes.

By the way, the `as` casts I've added are ugly and probably make it less readable, but I think on balance preferable to avoid `// @ts-expect-error` comments where possible. So annoying that TS doesn't allow type-asserting side effects of functions. It seems designed to guide you *away* from performant patterns, and make you introduce extra temp variables everywhere if you want to type cast.
